packages/pwa-buildpack/src/WebpackTools/plugins/makeMagentoRootComponentsPlugin/index.js

```diff
@@ -0,0 +1,126 @@
+                    deep: true,
+                    filter: /m?[jt]s$/
+                });
+                await Promise.all(
+                    rootComponentFiles.map(async rootComponentFile => {
+                        const rootComponentSource = await readFile(
+                            rootComponentFile,
+                            'utf8'
+                        );
+                        const { directives = [], errors } = directiveParser(
+                            rootComponentSource
+                        );
+                        if (errors.length) {
+                            // for now, errors just mean no directive was found
+                            return;
+                        }
+                        const rootComponentDirectives = directives.filter(
+                            d => d.type === 'RootComponent'
+                        );
+
+                        if (rootComponentDirectives.length === 0) {
+                            return;
+                        }
+
+                        if (rootComponentDirectives.length > 1) {
+                            console.warn(
+                                `Found more than 1 RootComponent Directive in ${rootComponentFile}. Only the first will be used`
+                            );
+                        }
+
+                        const {
+                            pageTypes,
+                            variant
+                        } = rootComponentDirectives[0];
+
+                        if (!pageTypes || pageTypes.length === 0) {
+                            console.warn(
+                                `No pageTypes specified for RootComponent ${rootComponentFile}. RootComponent will never be used.`
+                            );
+                        } else {
+                            pageTypes.forEach(pageType => {
+                                const key = toRootComponentMapKey(
+                                    pageType,
+                                    variant
+                                );
+                                importerSources[
+                                    key
+                                ] = `() => import(/* webpackChunkName: "${key}" */'${relative(
+                                    context,
+                                    rootComponentFile
+                                )}')`;
+                            });
+                        }
+                    })
+                );
+                return importerSources;
+            },
+            Promise.resolve({})
+        );
+
+        this.contents = `
+const rootComponentsMap = {
+${Object.entries(rootComponentImporters)
+            .map(entry => entry.join(':'))
+            .join(',\n')}
+};
+const key = ${toRootComponentMapKey.toString()};
+export default function fetchRootComponent(type, variant = 'default') {
+    return rootComponentsMap[key(type, variant)]().then(m => m.default || m);
+};
+`;
+    }
+}
```

Comment:
You know, I actually kind of like it, but maybe I've been looking at it too long. This basic idea of building a "dynamic import dispatch" file is promising to me, because it lets us set up a big registry of dynamic imports which Webpack can detect and chunk, entirely through filesystem analysis, but we could still set up an arbitrary API for other modules to trigger the dynamic import. I think this kind of thing could be the basis for future extensibility concepts. Though I suppose we'd build it with proper codegen.

Reply:
Okay. Mostly, I just don't want to have to maintain code that's written in a string. This plugin has a limited job, though, and it's doing it well. I think this will be fine.